jax/_src/numpy/ufuncs.py

```diff
@@ -732,10 +733,21 @@ def heaviside(x1: ArrayLike, x2: ArrayLike, /) -> Array:
 def hypot(x1: ArrayLike, x2: ArrayLike, /) -> Array:
   check_arraylike("hypot", x1, x2)
   x1, x2 = promote_dtypes_inexact(x1, x2)
-  x1 = lax.abs(x1)
-  x2 = lax.abs(x2)
+
+  # TODO(micky774): Promote to ValueError when deprecation is complete
+  # (began 2024-4-14).
+  if dtypes.issubdtype(x1.dtype, np.complexfloating):
+    warnings.warn(
+      "Passing complex-valued inputs to hypot is deprecated and will raise a "
+      "ValueError in the future. Please convert to real values first, such as "
+      "by using jnp.real or jnp.imag to take the real or imaginary components "
+      "respectively.",
+      DeprecationWarning, stacklevel=2)
+  x1, x2 = lax.abs(x1), lax.abs(x2)
+  idx_inf = lax.bitwise_or(isposinf(x1), isposinf(x2))
   x1, x2 = maximum(x1, x2), minimum(x1, x2)
-  return lax.select(x1 == 0, x1, x1 * lax.sqrt(1 + lax.square(lax.div(x2, lax.select(x1 == 0, lax._ones(x1), x1)))))
+  x = _where(x1 == 0, x1, x1 * lax.sqrt(1 + lax.square(lax.div(x2, _where(x1 == 0, lax._ones(x1), x1)))))
+  return _where(idx_inf, _lax_const(x, np.inf), x)
```

Comment:
I'm not sure why broadcasting fails when including the calculation of `idx_inf`.

Reply:
Unfortunately `maximum(np.nan, np.inf) = np.nan` so those cases will lead to incorrect output.

Reply:
I assume the expected behavior is `hypot(nan, inf)` -> `nan`. From reading the code, `x1` and `x2` will be both `nan` after `x1, x2 = maximum(x1, x2), minimum(x1, x2)`. So will be `x` and the returned value. Perhaps I am missing something here..

Reply:
Actually the array API standard specifies `hypot(nan, +-inf) = inf`

Reply:
I had the same assumption as @pearu – I don't understand the reasoning behind suppressing the propagation of NaN values here.

Reply:
Just for reference, apparently it is specified by IEEE-754, see [page 62 here](http://deb.scienceontheweb.net/Informatica/lezioni/IEEE754_ver2019.pdf)

Reply:
Thanks for the reference!

Reply:
To resolve this I've changed `check_arraylike, promote_dtypes_inexact-->promote_args_inexact` which also handles broadcasting.